### PR TITLE
LSM-based aggregation

### PR DESCRIPTION
**Pull request by carsonip**
## Motivation/summary

Replace existing aggregation with apm-aggregation, the LSM-based aggregator.

Tasks:
- [x] Pass configured limits to apm-aggregation
- [x] Fix various test failures
- [x] ~~Storage configuration~~ In-memory FS used by pebble
- [x] Benchmarks
- [x] Remove old aggregation code
~~- [ ] aggregator Close should log if there's an error (#11355)~~

## Checklist

~~(no user facing changes)- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)~~
~~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~~
~~- [ ] Documentation has been updated~~

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## How to test these changes

- Ensure that APM UI works as usual

### CODE REVIEW

**Comment by carsonip** [comment]
## Progress update

I've been working on understanding and optimizing the performance of LSM PoC. Performance improvements are WIP in apm-aggregation repo. But here's a glimpse of how LSM PoC fares against main:
```
                           │ /home/carson/projects/apm-server/apmbench-out/bench-main-2-merge.txt │ /home/carson/projects/apm-server/apmbench-out/bench-after-alloc-after-vtproto-2-merge.txt │
                           │                                sec/op                                │                              sec/op                                vs base                │
1000TransactionsSerial-512                                                           13.45m ± ∞ ¹                                                        21.76m ± ∞ ¹       ~ (p=0.333 n=2) ²
AgentAll-512                                                                         393.2m ± ∞ ¹                                                        418.0m ± ∞ ¹       ~ (p=0.333 n=2) ²
AgentGo-512                                                                          104.2m ± ∞ ¹                                                        103.7m ± ∞ ¹       ~ (p=0.667 n=2) ²
AgentNodeJS-512                                                                      51.13m ± ∞ ¹                                                        50.23m ± ∞ ¹       ~ (p=1.000 n=2) ²
AgentPython-512                                                                      162.8m ± ∞ ¹                                                        139.4m ± ∞ ¹       ~ (p=0.333 n=2) ²
AgentRuby-512                                                                        89.96m ± ∞ ¹                                                        88.98m ± ∞ ¹       ~ (p=0.667 n=2) ²
geomean                                                                              86.29m                                                              91.51m        +6.05%
¹ need >= 6 samples for confidence interval at level 0.95
² need >= 4 samples to detect a difference at alpha level 0.05

                           │ /home/carson/projects/apm-server/apmbench-out/bench-main-2-merge.txt │ /home/carson/projects/apm-server/apmbench-out/bench-after-alloc-after-vtproto-2-merge.txt │
                           │                              events/sec                              │                            events/sec                              vs base                │
1000TransactionsSerial-512                                                           74.44k ± ∞ ¹                                                        45.95k ± ∞ ¹       ~ (p=0.333 n=2) ²
AgentAll-512                                                                         45.84k ± ∞ ¹                                                        43.21k ± ∞ ¹       ~ (p=0.333 n=2) ²
AgentGo-512                                                                          50.22k ± ∞ ¹                                                        50.50k ± ∞ ¹       ~ (p=0.667 n=2) ²
AgentNodeJS-512                                                                      39.47k ± ∞ ¹                                                        40.19k ± ∞ ¹       ~ (p=1.000 n=2) ²
AgentPython-512                                                                      43.19k ± ∞ ¹                                                        50.38k ± ∞ ¹       ~ (p=0.333 n=2) ²
AgentRuby-512                                                                        41.73k ± ∞ ¹                                                        42.21k ± ∞ ¹       ~ (p=0.667 n=2) ²
geomean                                                                              47.97k                                                              45.24k        -5.70%
¹ need >= 6 samples for confidence interval at level 0.95
² need >= 4 samples to detect a difference at alpha level 0.05

                           │ /home/carson/projects/apm-server/apmbench-out/bench-main-2-merge.txt │ /home/carson/projects/apm-server/apmbench-out/bench-after-alloc-after-vtproto-2-merge.txt │
                           │                                 B/op                                 │                              B/op                                vs base                  │
1000TransactionsSerial-512                                                          5.667Mi ± ∞ ¹                                                    20.012Mi ± ∞ ¹         ~ (p=0.333 n=2) ²
AgentAll-512                                                                        251.7Mi ± ∞ ¹                                                     492.3Mi ± ∞ ¹         ~ (p=0.333 n=2) ²
AgentGo-512                                                                         63.21Mi ± ∞ ¹                                                    120.73Mi ± ∞ ¹         ~ (p=0.333 n=2) ²
AgentNodeJS-512                                                                     30.88Mi ± ∞ ¹                                                     59.81Mi ± ∞ ¹         ~ (p=0.333 n=2) ²
AgentPython-512                                                                     101.3Mi ± ∞ ¹                                                     200.2Mi ± ∞ ¹         ~ (p=0.333 n=2) ²
AgentRuby-512                                                                       59.99Mi ± ∞ ¹                                                    107.15Mi ± ∞ ¹         ~ (p=0.333 n=2) ²
geomean                                                                             50.67Mi                                                           107.3Mi        +111.75%
¹ need >= 6 samples for confidence interval at level 0.95
² need >= 4 samples to detect a difference at alpha level 0.05

                           │ /home/carson/projects/apm-server/apmbench-out/bench-main-2-merge.txt │ /home/carson/projects/apm-server/apmbench-out/bench-after-alloc-after-vtproto-2-merge.txt │
                           │                              allocs/op                               │                            allocs/op                              vs base                 │
1000TransactionsSerial-512                                                           73.25k ± ∞ ¹                                                      162.97k ± ∞ ¹        ~ (p=0.333 n=2) ²
AgentAll-512                                                                         3.964M ± ∞ ¹                                                       5.090M ± ∞ ¹        ~ (p=0.333 n=2) ²
AgentGo-512                                                                          854.1k ± ∞ ¹                                                      1151.5k ± ∞ ¹        ~ (p=0.333 n=2) ²
AgentNodeJS-512                                                                      515.0k ± ∞ ¹                                                       659.8k ± ∞ ¹        ~ (p=0.333 n=2) ²
AgentPython-512                                                                      1.700M ± ∞ ¹                                                       2.103M ± ∞ ¹        ~ (p=0.333 n=2) ²
AgentRuby-512                                                                        940.8k ± ∞ ¹                                                      1182.0k ± ∞ ¹        ~ (p=0.333 n=2) ²
geomean                                                                              767.4k                                                             1.078M        +40.42%
¹ need >= 6 samples for confidence interval at level 0.95
² need >= 4 samples to detect a difference at alpha level 0.05
```

The numbers are quite unstable since I ran the benchmarks on my laptop. But other than 1000TransactionsSerial (similar to 1000Transactions but runs in serial instead of in parallel), all the agent benchmarks seem to be fairly similar.

File: systemtest/approvals/TestServiceSummaryMetricsAggregationOverflow.approved.json [hunk 1]
```diff
@@ -139,7 +139,7 @@
                 "name": "metric"
             },
             "service": {
-                "environment": "dev",
+                "environment": "prod",
```
Comment (carsonip) [inline systemtest/approvals/TestServiceSummaryMetricsAggregationOverflow.approved.json:142]: [for reviewer] service summary metrics map iteration order is non-deterministic. This change is a result of change in underlying aggregation implementation.

File: x-pack/apm-server/main.go [hunk 2]
```diff
@@ -100,85 +90,19 @@ type processor interface {
...
-		HDRHistogramSignificantFigures: args.Config.Aggregation.Transactions.HDRHistogramSignificantFigures,
-	})
+	processors := make([]namedProcessor, 0, 2)
+
+	name := "LSM aggregator"
+	agg, err := aggregation.NewAggregator(context.TODO(),
+		args.Config.Aggregation.MaxServices,
+		args.Config.Aggregation.Transactions.MaxGroups,
+		args.Config.Aggregation.ServiceTransactions.MaxGroups,
+		args.Config.Aggregation.ServiceDestinations.MaxGroups,
```
Comment (carsonip) [inline x-pack/apm-server/main.go:100]: Service destination still has a fixed default of 10k max groups. It does not change according to memory available. Do we want to change that?

File: x-pack/apm-server/aggregation/lsm.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+	baseaggregator, err := aggregators.New(
+		aggregators.WithLimits(aggregators.Limits{
+			MaxSpanGroups:                         maxSpanGroups,
+			MaxSpanGroupsPerService:               max(maxSpanGroups/10, 1),
+			MaxTransactionGroups:                  maxTxGroups,
+			MaxTransactionGroupsPerService:        max(maxTxGroups/10, 1),
+			MaxServiceTransactionGroups:           maxSvcTxGroups,
+			MaxServiceTransactionGroupsPerService: max(maxSvcTxGroups/10, 1),
+			MaxServiceInstanceGroupsPerService:    max(maxSvcs/10, 1),
```
Comment (carsonip) [inline x-pack/apm-server/aggregation/lsm.go:43]: We have not yet settled on whether to keep service instance or to revert it. In any case, it should not be blocking this PR. Meanwhile, is `maxSvcs/10` a sane enough default for MaxServiceInstanceGroupsPerService?

**Comment by carsonip** [comment]
Performance is about as good as main branch as of 03a55a7. Bear in mind that these are the numbers from the specific github actions setup and smaller/larger instances may report a different delta. Also, the variance is large, that the difference between a good run and a bad run may be huge.

In github action benchmark run 91 (main) vs 93 (lsm branch):
```
                           │ github/run91.txt │        github/run93-lsm.txt        │
                           │      sec/op      │    sec/op     vs base              │
10000AggregationGroups-512       902.1m ±  4%   914.1m ± 54%       ~ (p=0.937 n=6)
1000Transactions-512             32.50m ± 53%   42.94m ± 60%       ~ (p=0.065 n=6)
AgentAll-512                      1.670 ± 13%    1.800 ± 16%       ~ (p=0.180 n=6)
AgentGo-512                      354.2m ± 24%   358.0m ± 34%       ~ (p=0.937 n=6)
AgentNodeJS-512                  278.3m ± 24%   282.6m ± 28%       ~ (p=1.000 n=6)
AgentPython-512                  801.0m ± 31%   734.6m ± 24%       ~ (p=0.589 n=6)
AgentRuby-512                    444.0m ± 30%   384.6m ± 25%       ~ (p=0.310 n=6)
geomean                          402.7m         412.1m        +2.35%

                           │  github/run91.txt   │            github/run93-lsm.txt             │
                           │ error_responses/sec │ error_responses/sec  vs base                │
10000AggregationGroups-512          0.000 ± 0%              0.000 ± 0%       ~ (p=1.000 n=6) ¹
1000Transactions-512                0.000 ± 0%              0.000 ± 0%       ~ (p=1.000 n=6) ¹
AgentAll-512                        0.000 ± 0%              0.000 ± 0%       ~ (p=1.000 n=6) ¹
AgentGo-512                         0.000 ± 0%              0.000 ± 0%       ~ (p=1.000 n=6) ¹
AgentNodeJS-512                     0.000 ± 0%              0.000 ± 0%       ~ (p=1.000 n=6) ¹
AgentPython-512                     0.000 ± 0%              0.000 ± 0%       ~ (p=1.000 n=6) ¹
AgentRuby-512                       0.000 ± 0%              0.000 ± 0%       ~ (p=1.000 n=6) ¹
geomean                                        ²                        +0.00%               ²
¹ all samples are equal
² summaries must be >0 to compute geomean

                           │ github/run91.txt │        github/run93-lsm.txt         │
                           │    errors/sec    │ errors/sec   vs base                │
10000AggregationGroups-512      0.000 ±  0%     0.000 ±  0%       ~ (p=1.000 n=6) ¹
1000Transactions-512            0.000 ±  0%     0.000 ±  0%       ~ (p=1.000 n=6) ¹
AgentAll-512                    215.6 ± 15%     200.4 ± 19%       ~ (p=0.180 n=6)
AgentGo-512                     242.1 ± 28%     269.8 ± 42%       ~ (p=0.394 n=6)
AgentNodeJS-512                 176.5 ± 31%     173.4 ± 39%       ~ (p=1.000 n=6)
AgentPython-512                 90.21 ± 45%     98.03 ± 32%       ~ (p=0.589 n=6)
AgentRuby-512                   304.9 ± 41%     350.0 ± 33%       ~ (p=0.310 n=6)
geomean                                     ²                +3.47%               ²
¹ all samples are equal
² summaries must be >0 to compute geomean

                           │ github/run91.txt │        github/run93-lsm.txt        │
                           │    events/sec    │  events/sec   vs base              │
10000AggregationGroups-512       18.55k ±  3%   17.67k ± 34%       ~ (p=0.132 n=6)
1000Transactions-512             30.97k ± 35%   23.40k ± 37%       ~ (p=0.065 n=6)
AgentAll-512                     10.80k ± 15%   10.04k ± 19%       ~ (p=0.180 n=6)
AgentGo-512                      11.73k ± 27%   13.24k ± 44%       ~ (p=0.310 n=6)
AgentNodeJS-512                  7.249k ± 31%   7.123k ± 39%       ~ (p=1.000 n=6)
AgentPython-512                  8.797k ± 45%   9.559k ± 32%       ~ (p=0.589 n=6)
AgentRuby-512                    8.541k ± 41%   9.807k ± 33%       ~ (p=0.310 n=6)
geomean                          12.17k         12.04k        -1.09%

                           │ github/run91.txt │         github/run93-lsm.txt          │
                           │    gc_cycles     │  gc_cycles    vs base                 │
10000AggregationGroups-512        19.50 ± 18%   323.50 ±  8%  +1558.97% (p=0.002 n=6)
1000Transactions-512              51.50 ± 34%   140.00 ± 19%   +171.84% (p=0.002 n=6)
AgentAll-512                      67.50 ±  7%   112.00 ±  9%    +65.93% (p=0.002 n=6)
AgentGo-512                       50.50 ± 21%    66.00 ± 20%    +30.69% (p=0.004 n=6)
AgentNodeJS-512                   51.00 ± 49%    93.00 ± 49%    +82.35% (p=0.022 n=6)
AgentPython-512                   42.50 ± 41%    56.50 ± 36%          ~ (p=0.065 n=6)
AgentRuby-512                     53.50 ± 51%    68.00 ± 31%          ~ (p=0.132 n=6)
geomean                           45.39          102.6         +126.02%

                           │ github/run91.txt │         github/run93-lsm.txt         │
                           │  max_goroutines  │ max_goroutines  vs base              │
10000AggregationGroups-512        173.0 ±  8%      169.5 ±  4%       ~ (p=0.457 n=6)
1000Transactions-512              715.0 ± 35%      797.5 ± 57%       ~ (p=0.818 n=6)
AgentAll-512                      254.5 ±  7%      254.5 ±  4%       ~ (p=0.732 n=6)
AgentGo-512                       637.0 ±  1%      502.5 ± 28%       ~ (p=0.177 n=6)
AgentNodeJS-512                   640.0 ± 17%      630.0 ± 13%       ~ (p=0.457 n=6)
AgentPython-512                   379.0 ± 33%      342.5 ± 25%       ~ (p=0.513 n=6)
AgentRuby-512                     604.0 ± 33%      623.0 ± 32%       ~ (p=0.589 n=6)
geomean                           434.8            420.5        -3.29%

                           │ github/run91.txt │         github/run93-lsm.txt          │
                           │  max_heap_alloc  │ max_heap_alloc  vs base               │
10000AggregationGroups-512      2210.7M ±  1%     339.2M ± 27%  -84.66% (p=0.002 n=6)
1000Transactions-512            1457.4M ± 37%     759.3M ± 43%  -47.91% (p=0.004 n=6)
AgentAll-512                    1105.3M ±  5%     847.6M ±  6%  -23.32% (p=0.002 n=6)
AgentGo-512                     1280.1M ± 12%     903.0M ± 24%  -29.46% (p=0.002 n=6)
AgentNodeJS-512                 1178.5M ±  6%     940.5M ± 10%  -20.19% (p=0.002 n=6)
AgentPython-512                  1.351G ±  5%     1.048G ± 14%  -22.41% (p=0.002 n=6)
AgentRuby-512                    1.216G ± 16%     1.146G ± 18%        ~ (p=0.180 n=6)
geomean                          1.365G           806.9M        -40.88%

                           │ github/run91.txt │          github/run93-lsm.txt           │
                           │ max_heap_objects │ max_heap_objects  vs base               │
10000AggregationGroups-512      12.109M ±  2%       2.011M ± 27%  -83.39% (p=0.002 n=6)
1000Transactions-512             7.992M ± 25%       3.441M ± 56%  -56.94% (p=0.002 n=6)
AgentAll-512                     7.395M ±  5%       5.082M ±  4%  -31.28% (p=0.002 n=6)
AgentGo-512                      7.109M ± 10%       4.684M ± 25%  -34.11% (p=0.002 n=6)
AgentNodeJS-512                  8.271M ±  7%       5.993M ±  7%  -27.54% (p=0.002 n=6)
AgentPython-512                  9.375M ±  7%       6.446M ± 18%  -31.24% (p=0.002 n=6)
AgentRuby-512                    8.632M ± 17%       7.493M ± 19%  -13.20% (p=0.026 n=6)
geomean                          8.574M             4.660M        -45.65%

                           │ github/run91.txt │        github/run93-lsm.txt         │
                           │     max_rss      │   max_rss     vs base               │
10000AggregationGroups-512      2080.8M ±  0%   455.2M ± 16%  -78.12% (p=0.002 n=6)
1000Transactions-512            1959.6M ± 49%   918.9M ± 37%  -53.11% (p=0.009 n=6)
AgentAll-512                    1115.6M ±  5%   988.9M ±  3%  -11.36% (p=0.002 n=6)
AgentGo-512                      1.295G ± 12%   1.056G ± 24%  -18.44% (p=0.026 n=6)
AgentNodeJS-512                  1.186G ±  7%   1.062G ±  5%  -10.49% (p=0.002 n=6)
AgentPython-512                  1.358G ±  6%   1.209G ± 12%  -10.92% (p=0.004 n=6)
AgentRuby-512                    1.203G ± 15%   1.269G ± 15%        ~ (p=0.485 n=6)
geomean                          1.416G         952.7M        -32.72%

                           │    github/run91.txt     │              github/run93-lsm.txt              │
                           │ mean_available_indexers │ mean_available_indexers  vs base               │
10000AggregationGroups-512               14.52 ± 10%               15.28 ± 14%   +5.16% (p=0.041 n=6)
1000Transactions-512                     10.53 ± 35%               13.61 ± 22%  +29.20% (p=0.026 n=6)
AgentAll-512                             15.57 ±  6%               16.41 ±  5%        ~ (p=0.093 n=6)
AgentGo-512                              15.88 ± 10%               15.38 ± 15%        ~ (p=0.699 n=6)
AgentNodeJS-512                          17.05 ±  8%               17.34 ±  9%        ~ (p=0.221 n=6)
AgentPython-512                          17.22 ±  8%               17.04 ±  7%        ~ (p=0.818 n=6)
AgentRuby-512                            16.97 ±  9%               16.51 ±  8%        ~ (p=0.485 n=6)
geomean                                  15.21                     15.89         +4.47%

                           │ github/run91.txt │         github/run93-lsm.txt         │
                           │   metrics/sec    │  metrics/sec   vs base               │
10000AggregationGroups-512       1.558k ±  8%   1.192k ±  21%  -23.52% (p=0.002 n=6)
1000Transactions-512              28.50 ± 11%    27.64 ± 190%        ~ (p=0.818 n=6)
AgentAll-512                     1.656k ± 15%   1.538k ±  19%        ~ (p=0.180 n=6)
AgentGo-512                       590.4 ± 29%    642.4 ±  37%        ~ (p=0.556 n=6)
AgentNodeJS-512                   431.5 ± 31%    424.1 ±  39%        ~ (p=1.000 n=6)
AgentPython-512                  2.655k ± 45%   2.883k ±  32%        ~ (p=0.589 n=6)
AgentRuby-512                     627.2 ± 41%    720.8 ±  33%        ~ (p=0.310 n=6)
geomean                           609.4          602.1          -1.20%

                           │ github/run91.txt │         github/run93-lsm.txt         │
                           │    spans/sec     │  spans/sec    vs base                │
10000AggregationGroups-512     8.900k ±  3%     8.723k ± 35%       ~ (p=0.818 n=6)
1000Transactions-512            0.000 ±  0%      0.000 ±  0%       ~ (p=1.000 n=6) ¹
AgentAll-512                   6.468k ± 15%     6.012k ± 19%       ~ (p=0.180 n=6)
AgentGo-512                    7.715k ± 27%     8.719k ± 45%       ~ (p=0.310 n=6)
AgentNodeJS-512                3.821k ± 31%     3.755k ± 39%       ~ (p=1.000 n=6)
AgentPython-512                5.111k ± 45%     5.553k ± 32%       ~ (p=0.589 n=6)
AgentRuby-512                  5.024k ± 41%     5.769k ± 33%       ~ (p=0.310 n=6)
geomean                                     ²                 +3.38%               ²
¹ all samples are equal
² summaries must be >0 to compute geomean

                           │ github/run91.txt │        github/run93-lsm.txt        │
                           │     txs/sec      │   txs/sec     vs base              │
10000AggregationGroups-512       8.099k ±  2%   7.781k ± 35%       ~ (p=0.372 n=6)
1000Transactions-512             30.94k ± 35%   23.37k ± 38%       ~ (p=0.065 n=6)
AgentAll-512                     2.459k ± 15%   2.286k ± 19%       ~ (p=0.180 n=6)
AgentGo-512                      3.183k ± 27%   3.611k ± 45%       ~ (p=0.310 n=6)
AgentNodeJS-512                  2.820k ± 31%   2.771k ± 39%       ~ (p=1.000 n=6)
AgentPython-512                   942.2 ± 45%   1023.5 ± 32%       ~ (p=0.589 n=6)
AgentRuby-512                    2.585k ± 41%   2.967k ± 34%       ~ (p=0.310 n=6)
geomean                          3.890k         3.854k        -0.91%

                           │ github/run91.txt │         github/run93-lsm.txt         │
                           │       B/op       │     B/op       vs base               │
10000AggregationGroups-512      115.2Mi ±  6%   223.8Mi ±  3%  +94.22% (p=0.002 n=6)
1000Transactions-512            6.115Mi ± 21%   9.694Mi ± 16%  +58.53% (p=0.002 n=6)
AgentAll-512                    213.5Mi ±  0%   243.7Mi ±  0%  +14.14% (p=0.002 n=6)
AgentGo-512                     42.66Mi ±  5%   59.37Mi ± 20%  +39.15% (p=0.002 n=6)
AgentNodeJS-512                 26.38Mi ±  1%   31.00Mi ±  2%  +17.54% (p=0.002 n=6)
AgentPython-512                 86.35Mi ±  2%   98.49Mi ±  1%  +14.06% (p=0.002 n=6)
AgentRuby-512                   50.67Mi ±  1%   56.68Mi ±  1%  +11.87% (p=0.002 n=6)
geomean                         49.62Mi         65.96Mi        +32.93%

                           │ github/run91.txt │         github/run93-lsm.txt         │
                           │    allocs/op     │   allocs/op    vs base               │
10000AggregationGroups-512       1.904M ±  4%    2.936M ±  2%  +54.21% (p=0.002 n=6)
1000Transactions-512             89.13k ± 12%   135.53k ± 10%  +52.07% (p=0.002 n=6)
AgentAll-512                     3.894M ±  0%    4.268M ±  0%   +9.62% (p=0.002 n=6)
AgentGo-512                      665.8k ±  5%    918.1k ± 20%  +37.90% (p=0.002 n=6)
AgentNodeJS-512                  502.1k ±  1%    555.8k ±  1%  +10.70% (p=0.002 n=6)
AgentPython-512                  1.668M ±  1%    1.797M ±  0%   +7.74% (p=0.002 n=6)
AgentRuby-512                    937.9k ±  1%   1017.0k ±  1%   +8.43% (p=0.002 n=6)
geomean                          859.1k          1.068M        +24.30%
```

File: x-pack/apm-server/aggregation/lsm.go (hunk 3 above)
Comment (axw) [reply x-pack/apm-server/aggregation/lsm.go:43]: I think it's fine to come back to it in a followup, but soon. Let's go with `maxSvcs/10` for now and create an issue to finalise the decision.

File: x-pack/apm-server/main.go (hunk 2 above)
Comment (axw) [reply x-pack/apm-server/main.go:100]: I think we should minimise the number of changes. Let's aim to change the technical implementation details (i.e. memory -> LSM) and make improvements in a following release, so long as it's not a one way door.

File: systemtest/aggregation_test.go [hunk 4]
```diff
@@ -77,10 +71,6 @@ func TestTransactionAggregation(t *testing.T) {
 		espoll.TermQuery{Field: "processor.event", Value: "transaction"},
 	)
 
-	// Make sure apm-server.aggregation.txmetrics metrics are published. Metric values are unit tested.
-	doc := getBeatsMonitoringStats(t, srv, nil)
-	assert.True(t, gjson.GetBytes(doc.RawSource, "beats_stats.metrics.apm-server.aggregation.txmetrics").Exists())
```
Comment (axw) [inline systemtest/aggregation_test.go:82]: We probably discussed this in the past and I forgot. We're not relying on this in any part of stack monitoring, supportability dashboards, or product telemetry, right?

File: x-pack/apm-server/aggregation/lsm.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type Aggregator struct {
+	logger         *zap.Logger
+	baseaggregator *aggregators.Aggregator
+}
+
+// NewAggregator returns a new instance of aggregator.
+func NewAggregator(
```
Comment (axw) [inline x-pack/apm-server/aggregation/lsm.go:27]:
```suggestion
// New returns a new Aggregator.
func New(
```

File: x-pack/apm-server/aggregation/lsm.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	ctx context.Context,
```
Comment (axw) [inline x-pack/apm-server/aggregation/lsm.go:28]:
```suggestion
```

File: x-pack/apm-server/main.go [hunk 7]
```diff
@@ -100,85 +90,19 @@ type processor interface {
...
-		MetricsInterval:                metricsInterval,
-		RollUpIntervals:                rollUpMetricsIntervals,
-		MaxTransactionGroupsPerService: int(math.Ceil(0.1 * float64(args.Config.Aggregation.Transactions.MaxTransactionGroups))),
-		MaxServices:                    args.Config.Aggregation.Transactions.MaxServices,
-		HDRHistogramSignificantFigures: args.Config.Aggregation.Transactions.HDRHistogramSignificantFigures,
-	})
+	processors := make([]namedProcessor, 0, 2)
+
+	name := "LSM aggregator"
+	agg, err := aggregation.NewAggregator(context.TODO(),
```
Comment (axw) [inline x-pack/apm-server/main.go:96]:
```suggestion
	agg, err := aggregation.New(
```

File: x-pack/apm-server/main.go [hunk 8]
```diff
@@ -100,85 +90,19 @@ type processor interface {
...
-	agg, err := txmetrics.NewAggregator(txmetrics.AggregatorConfig{
-		BatchProcessor:                 args.BatchProcessor,
-		MaxTransactionGroups:           args.Config.Aggregation.Transactions.MaxTransactionGroups,
-		MetricsInterval:                metricsInterval,
-		RollUpIntervals:                rollUpMetricsIntervals,
-		MaxTransactionGroupsPerService: int(math.Ceil(0.1 * float64(args.Config.Aggregation.Transactions.MaxTransactionGroups))),
-		MaxServices:                    args.Config.Aggregation.Transactions.MaxServices,
-		HDRHistogramSignificantFigures: args.Config.Aggregation.Transactions.HDRHistogramSignificantFigures,
-	})
+	processors := make([]namedProcessor, 0, 2)
```
Comment (axw) [inline x-pack/apm-server/main.go:93]:
```suggestion
	var processors []namedProcessor
```

This is not a hot code path, so let's simplify this

File: x-pack/apm-server/aggregation/lsm.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+	agg := &Aggregator{
+		logger:         zapLogger,
+		baseaggregator: baseaggregator,
+	}
+	return agg, err
+}
+
+// Run runs all the components of aggregator.
+func (a *Aggregator) Run() error {
+	return a.baseaggregator.Run(context.TODO())
```
Comment (axw) [inline x-pack/apm-server/aggregation/lsm.go:66]:
```suggestion
	return a.baseaggregator.Run(context.Background())
```

Also should we be checking that the result is not `ErrAggregatorClosed`? IIANM that's an expected result, if Close is called.

**Review by axw — APPROVED** [review]
LGTM! Just a few minor things.

File: systemtest/aggregation_test.go (hunk 4 above)
Comment (carsonip) [reply systemtest/aggregation_test.go:82]: There is one internal dashboard tracking overflows. We will need apm-aggregation to emit metrics around overflows, before we create a new dashboard for it.

Comment (carsonip) [reply systemtest/aggregation_test.go:82]: Will do in https://github.com/elastic/apm-aggregation/issues/75

File: changelogs/head.asciidoc [hunk 10]
```diff
@@ -21,3 +21,9 @@ https://github.com/elastic/apm-server/compare/8.9\...main[View commits]
 - Add a self-instrumentation transaction to the agent config periodic refresh action. {pull}11129[11129]
 - Stop dropping metadata fields from span documents. {pull}11089[11089]
 - Add permissions to reroute events in the integration package. {pull}11168[11168]
+
+[float]
+==== Aggregation improvements
+- Replace aggregation implementation {pull}11117[11117]
+- Add `service.language.name` to service destination metrics {pull}11117[11117]
+- Modify per-service transaction groups limit to consider more than service.name; Add per-service service destination groups limit and per-service service transaction groups limit {pull}11117[11117]
```
Comment (carsonip) [inline changelogs/head.asciidoc:29]: @simitt WDYT about the changelog?

**Review by kruskall — APPROVED** [review]

File: changelogs/head.asciidoc (hunk 10 above)
Comment (kruskall) [reply changelogs/head.asciidoc:29]: my 2c: I think we should improve the first message to briefly mention what it means for the user. At the very least we should mention the significantly reduced memory cost for aggregation by moving from in-memory to the on-disk LSM based approach.

Comment (carsonip) [reply changelogs/head.asciidoc:29]: I wanted to do so, but struggled to word it properly :facepalm:

Comment (carsonip) [reply changelogs/head.asciidoc]: Updated

File: dev_docs/trace_metrics.md [hunk 11]
```diff
@@ -15,26 +15,19 @@ As transactions are observed by APM Server, it groups them according to various
...
 of services tracked, the number of transaction groups tracked, as well as number
-of groups tracked per service.
-
-By default, the limits are 1,000 services per GB of memory, 5,000 transaction groups
-per GB of memory. When transaction group latency histograms are indexed, the groups
-are reset, enabling a different set of groups to be recorded.
-The per-service limit is 10% of the global limit. For example, for a 2GB APM Server,
-the limits are 2,000 services, 10,000 transaction groups, and for each service,
-there can be a maximum of 1,000 unique transaction groups.
+of groups tracked per service. See docs for limits.
```
Comment (carsonip) [inline dev_docs/trace_metrics.md:23]: Avoiding documenting the limits twice, as they are clearly documented in docs/data-model.asciidoc

Comment (axw) [inline dev_docs/trace_metrics.md:23]:
```suggestion
of groups tracked per service. See [docs](https://www.elastic.co/guide/en/apm/guide/current/data-model-metrics.html#_aggregated_metrics_limits_and_overflows) for limits.
```

**Comment by carsonip** [comment]
@kruskall do you mind leaving a note about where we did not return vt objects, or alternatively create an issue to do so? I'm very interested to know where we missed it.

**Pull request merged**